src/pdf_parser/pdf_utils/disambiguate_layout.py

```diff
@@ -0,0 +1,381 @@
+                                box_1,
+                                box_2,
+                                direction=reduction_direction,
+                                min_overlapping_pixels_horizontal=min_overlapping_pixels_horizontal,
+                            )
+                        else:
+                            rect_1, rect_2 = reduce_overlapping_boxes(
+                                box_2,
+                                box_1,
+                                direction=reduction_direction,
+                                min_overlapping_pixels_horizontal=min_overlapping_pixels_horizontal,
+                            )
+                    box_1 = rect_1
+                    box_2 = rect_2
+    return blocks
+
+
+def unnest_boxes(layout: Layout, unnest_soft_margin: int = 15) -> Layout:
+    """
+    Loop through boxes, unnest them until there are no nested boxes left..
+
+    Args: layout: The layout to unnest.
+    unnest_soft_margin: The number of pixels to inflate each box by in each direction
+    when checking for containment (i.e. a soft margin).
+
+    Returns:
+        The unnested boxes.
+    """
+    if len(layout) == 0:
+        return layout
+    else:
+        # Add a soft-margin for the is_in function to allow for some leeway in the containment check.
+        soft_margin = {
+            "top": unnest_soft_margin,
+            "bottom": unnest_soft_margin,
+            "left": unnest_soft_margin,
+            "right": unnest_soft_margin,
+        }
+        # The loop checks each block for containment within other blocks.
+        # Contained blocks are removed if they have lower confidence scores than their parents;
+        # otherwise, the parent is removed. The process continues until there are no contained blocks.
+        # There are potentially nestings within nestings, hence the complicated loop.
+        # A recursion might be more elegant, leaving it as a TODO.
+        stop_cond = True
+        counter = 0  # num contained blocks in every run through of all pairs to calculate stop
+        # condition.
+        ixs_to_remove = []
+        while stop_cond:
+            for ix, box_1 in enumerate(layout):
+                for ix2, box_2 in enumerate(layout):
+                    if box_1 == box_2:
+                        continue
+                    else:
+                        if box_1.is_in(box_2, soft_margin):
+                            counter += 1
+                            # Remove the box the model is less confident about.
+                            if box_1.score > box_2.score:
+                                remove_ix = ix2
+                            else:
+                                remove_ix = ix
+                            ixs_to_remove.append(remove_ix)
+                # stop condition: no contained blocks
+                if counter == 0:
+                    stop_cond = False
+                counter = 0
+
```

Comment:
@stefl14 this is the infinite loop we found when deploying GA which isn't on main and you've put back into the code. how did it get here?